src/components/layers/InteractivePointLayer/index.tsx

```diff
@@ -0,0 +1,286 @@
+  const lastHoverId = useRef<string | number | null>(null);
+  const dragging = useRef<{
+    pointID: string | number;
+    offset: mapboxgl.Point;
+  } | null>(null);
+
+  // Flag set once the underlying point layer has been added to the map
+  const [pointLayerID, setPointLayerID] = useState<string | null>(null);
+  const { map } = useContext(MapboxContext);
+
+  // Set up event handlers
+  useEffect(() => {
+    if (!map || pointLayerID === null) return;
+    const handleClick: NativeMapEventHandler = (e) => {
+      const id =
+        getPointsSortedByDistance(e)?.filter((f) => {
+          if (f.id === null) return false;
+          const match = pointIndex[f.id];
+          return match && match.clickable;
+        })[0]?.id ?? null;
+      if (id === undefined || id === null) return;
+
+      onClick?.(id, e);
+    };
+    // Handle hover events and drag events
+    const handleMouseMove: NativeMapEventHandler = (e) => {
+      // Don't do anything if we're currently dragging a point
+      if (dragging.current !== null) return;
+      const sortedFeatures = getPointsSortedByDistance(e);
+      const closestHoverableID =
+        sortedFeatures?.filter((f) => {
+          if (f.id === null) return false;
+          const match = pointIndex[f.id];
+          return match && match.hoverable;
+        })[0]?.id ?? null;
+
+      // Only fire hover events if the feature beneath the pointer has changed
+      if (closestHoverableID !== lastHoverId.current) {
+        if (lastHoverId.current !== null) {
+          onHoverLeave?.(lastHoverId.current, e);
+        }
+
+        if (closestHoverableID !== null) onHoverEnter?.(closestHoverableID, e);
+        lastHoverId.current = closestHoverableID;
+      }
+    };
+
+    const handleDrag: NativeMapEventHandler<
+      mapboxgl.MapMouseEvent | mapboxgl.MapTouchEvent
+    > = (e) => {
+      if ((e.originalEvent as TouchEvent).touches?.length > 1) {
+        e.preventDefault();
+        return;
+      }
+      // Fire a drag event if currently dragging anything
+      if (dragging.current !== null) {
+        const pointerProjected = map.project(e.lngLat);
+        const offsetLngLat = map.unproject(
+          pointerProjected.sub(dragging.current.offset)
+        );
+        onDrag?.(
+          dragging.current.pointID,
+          { longitude: offsetLngLat.lng, latitude: offsetLngLat.lat },
+          dragging.current.offset,
+          e
+        );
+      }
+    };
+
+    const handleMouseDown: NativeMapEventHandler<
+      mapboxgl.MapMouseEvent | mapboxgl.MapTouchEvent
+    > = (e) => {
+      // Don't handle a touchstart event if we're already dragging something
+      if (e.type === "touchstart" && dragging.current !== null) {
+        return;
+      }
+      const closestFeature = getPointsSortedByDistance(e)?.filter((f) => {
+        if (f.id === null) return false;
+        const match = pointIndex[f.id];
+        return match && match.draggable;
+      })[0];
```

Comment:
I think you can skip part of a potentially large array here by using `find()` here instead, but maybe these arrays will never get that huge that it would matter...

Reply:
Good point— it might be cleaner to read that way, too. I'll update.